internal/common/ingest/batch.go

```diff
@@ -1,27 +1,22 @@
-package batch
+package ingest
 
 import (
 	"time"
 
-	"github.com/G-Research/armada/internal/pulsarutils"
-
 	"k8s.io/apimachinery/pkg/util/clock"
 )
 
-// Batch batches up events from a channel.  Batches are created whenever maxItems InstructionSets have been
+// Batch batches up events from a channel.  Batches are created whenever maxItems have been
 // received or maxTimeout has elapsed since the last batch was created (whichever occurs first).
-// This function has a lot in common with lookoutingester.batch.  Hopefully when generics become available we can
-// factor out most of the common code
-func Batch(values <-chan *pulsarutils.ConsumerMessage, maxItems int,
-	maxTimeout time.Duration, bufferSize int, clock clock.Clock,
-) chan []*pulsarutils.ConsumerMessage {
-	out := make(chan []*pulsarutils.ConsumerMessage, bufferSize)
+func Batch[T any](values <-chan T, maxItems int,
```

Comment:
Did you consider basing the batcher on a struct with a type parameter? Similar to
https://github.com/ninedraft/batch

The above batcher uses callbacks, which makes it a bit more generic too.

Why do we need to pass in a clock instead of using time functions?

Reply:
In this function, I think we can avoid the goto. Also, by passing in a context and not spawning a new goroutine inside the function, we can enable running the batcher in an errgroup. The caller would be responsible for spawning the goroutine (or not). That would also give the caller the choice of running the batcher async or sync.